V3.3.4 liberty-maven-plugin : maven plugin "classifier" valuation implies CWWKM2171E

Dear all,

When looseApplication=false, as soon as org.apache.maven.plugins:maven-**war**-plugin/maven-**ear**-plugin/maven-**ejb**-plugin is configured, in a superpom, with a **classifier** clause, the liberty-maven-plugin implies a **CWWKM2171E** on deploy.

**_investigation:_**
In liberty-plugin-config.xml, the **applicationFilename** is valued with "myapp-ear-0.0.1-SNAPSHOT.ear" instead of "myapp-ear-0.0.1-SNAPSHOT-liberty.ear", which is the obtained ear by maven-ear-plugin having "liberty" as valuation for maven plugin's **classifier** clause. Notabily, liberty maven plugin uses **applicationFilename=project.build.finalName** ; where finalName=${artifactId}-${version}, then is unable to retrieve the correct ear (==> classifer "-liberty" is missing from name).

**_tried workaround (unsuccess):_**
using "finalName" valuation containing : ${artifactId}-${version}-${target_server}
==> unsucess because maven-ear-plugin then creates a myapp-ear-0.0.1-SNAPSHOT-liberty-liberty.ear


>[INFO] --- maven-install-plugin:2.5.2:install (default-install) @ myapp-ear ---
[INFO] No primary artifact to install, installing attached artifacts instead.
[INFO] Installing C:\projects\prj-sbf-myapp\workspace\myapp\myapp-ear\pom.xml to C:\m2repo\com\s\b\myapp-ear\0.0.1-SNAPSHOT\myapp-ear-0.0.1-SNAPSHOT.pom
[INFO] Installing C:\projects\prj-sbf-myapp\workspace\myapp\myapp-ear\target\myapp-ear-0.0.1-SNAPSHOT-liberty.ear to C:\m2repo\com\s\b\myapp-ear\0.0.1-SNAPSHOT\myapp-ear-0.0.1-SNAPSHOT-liberty.ear
[INFO] 
[INFO] --- liberty-maven-plugin:3.3.4:deploy (thebas) @ myapp-ear ---
[INFO] CWWKM2102I: Utilisation de pre-installed assembly : F:\Desktop\b\wlp1.
[INFO] CWWKM2102I: Utilisation de serverName : theFASServer.
[INFO] CWWKM2102I: Utilisation de serverDirectory : F:\Desktop\b\wlp1\usr\servers\theFASServer.
[INFO] CWWKM2185I: The liberty-maven-plugin configuration parameter "appsDirectory" value defaults to "dropins".
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for SBAPP Root 0.0.1-SNAPSHOT:
[INFO] 
[INFO] SBAPP Root ......................................... SUCCESS [  0.869 s]
[INFO] SBAPP Service Contract ............................. SUCCESS [  1.126 s]
[INFO] SBAPP Resource Contract ............................ SUCCESS [  0.425 s]
[INFO] SBAPP Common ....................................... SUCCESS [  0.851 s]
[INFO] SBAPP Business ..................................... SUCCESS [  1.015 s]
[INFO] SBAPP Factory Settings ............................. SUCCESS [  1.811 s]
[INFO] SBAPP Batch ........................................ SUCCESS [  0.187 s]
[INFO] SBAPP EJB .......................................... SUCCESS [  0.819 s]
[INFO] SBAPP JAX-WS End Point ............................. SUCCESS [  0.451 s]
[INFO] SBAPP JAX-WS ....................................... SUCCESS [  0.483 s]
[INFO] SBAPP JAX-RS Client ................................ SUCCESS [  0.931 s]
[INFO] SBAPP JAX-RS ....................................... SUCCESS [  1.297 s]
[INFO] SBAPP JAX-WS Client ................................ SUCCESS [  0.121 s]
[INFO] SBAPP EAR .......................................... FAILURE [  1.642 s]
[INFO] SBAPP UI Web ....................................... SKIPPED
[INFO] SBAPP Web .......................................... SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  12.271 s
[INFO] Finished at: 2021-06-10T19:01:44+02:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal io.openliberty.tools:liberty-maven-plugin:3.3.4:deploy (thebas) on project myapp-ear: CWWKM2171E: The application for this project does not exist and cannot be installed. -> [Help 1]

@scottkurz, thank for your reply. I indeed did a mistake in the explanation within the readme.md of [earpomsample](https://github.com/placidetmuzo/earpomsample) : for `fake-project` it should be a `mvn install` (and not a `mvn package`). Here is the exact procedure to replay the error (windows command):

1) prerequisite : having OpenLiberty (full GA) installed in  `C:\workingarea\unzip\openliberty-21.0.0.6` folder. Else adapt the `installDirectory` clause of `liberty-maven-plugin` of `fake-project\fake-ear\pom.xml` file (L.63)

2) Start the OL server : `<path>\bin\server.bat run`

         C:\workingarea\unzip\openliberty-21.0.0.6>.\bin\server.bat run
         Listening for transport dt_socket at address: 7777
         Launching defaultServer (Open Liberty 21.0.0.6/wlp-1.0.53.cl210620210527-1900) on Eclipse OpenJ9 VM, version 1.8.0_292-b10 (en-us_FR)
         [AUDIT   ] CWWKE0001I: The server defaultServer has been launched.


3) install the fake-framework wich drive the classifier (as far as I understand you already did it) : 

        cd fake-framework-superpom
        mvn install
        ...
        [INFO] fake-framework-superpom                                            [pom]
        [INFO] fake-framework                                                     [pom]
        [INFO] fake-framework-common                                              [jar]

4) install the fake-project ; this will deploy the fake-project within running OL through liberty-maven-plugin of fake-project\fake-ear\pom.xml :

        cd fake-project
        mvn install
        ...
        [INFO] fake-project ....................................... SUCCESS [  0.203 s]
        [INFO] fake-jaxws ......................................... SUCCESS [  0.938 s]
        [INFO] fake-ejb ........................................... SUCCESS [  0.375 s]
        [INFO] fake-rest-jaxrs .................................... SUCCESS [  0.297 s]
        [INFO] fake-ear ........................................... FAILURE [ 47.189 s]
        [ERROR] Failed to execute goal io.openliberty.tools:liberty-maven-plugin:3.3.5-M2-SNAPSHOT:deploy (thebas) on project fake-ear: CWWKM2187E: Failed to deploy the fake-ear-0.0.1-SNAPSHOT-liberty application. The application start message was not found in the log file. -> [Help 1]

5) Step 4 implies the CWWKZ0117E also in OL console (continuation of step 2 just after dropins scan)  : 

        [AUDIT   ] CWWKZ0058I: Monitoring dropins for applications.
        [AUDIT   ] CWWKI0001I: The CORBA name server is now available at corbaloc:iiop:localhost:2809/NameService.
        [AUDIT   ] CWWKF0012I: The server installed the following features: [beanValidation-1.1, cdi-1.2, ejb-3.2, ejbHome-3.2, ejbLite-3.2, ejbPersistentTimer-3.2, ejbRemote-3.2, el-3.0, jaxb-2.2, jaxrs-2.0, jaxrsClient-2.0, jaxws-2.2, jca-1.7, jdbc-4.1, jndi-1.0, jpaContainer-2.1, json-1.0, mdb-3.2, servlet-3.1].
        [AUDIT   ] CWWKF0011I: The defaultServer server is ready to run a smarter planet. The defaultServer server started in 5.645 seconds.
        [AUDIT   ] CWWKG0016I: Starting server configuration update.
        [AUDIT   ] CWWKG0017I: The server configuration was successfully updated in 0.037 seconds.
        [ERROR   ] CWWKZ0117E: Application fake-ear-0.0.1-SNAPSHOT-liberty failed to locate module com.fake-fake-jaxws-0.0.1-SNAPSHOT-liberty.war of type web
        [ERROR   ] CWWKZ0117E: Application fake-ear-0.0.1-SNAPSHOT-liberty failed to locate module com.fake-fake-ejb-0.0.1-SNAPSHOT-liberty.jar of type ejb
        [ERROR   ] CWWKZ0117E: Application fake-ear-0.0.1-SNAPSHOT-liberty failed to locate module com.fake-fake-rest-jaxrs-0.0.1-SNAPSHOT-liberty.war of type web
        [ERROR   ] CWWKZ0124E: Application fake-ear-0.0.1-SNAPSHOT-liberty does not contain any modules.
        [AUDIT   ] CWWKZ0012I: The application fake-ear-0.0.1-SNAPSHOT-liberty was not started.

regards,

--- a/liberty-maven-plugin/src/it/ear-project-it/SampleEAR-classifier-app/pom.xml
+++ b/liberty-maven-plugin/src/it/ear-project-it/SampleEAR-classifier-app/pom.xml
@@ -1,0 +1,150 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+    xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+
+    <parent>
+        <groupId>io.openliberty.tools.it</groupId>
+        <artifactId>ear-project-it</artifactId>
+        <version>1.0-SNAPSHOT</version>
+    </parent>
+
+    <artifactId>SampleEAR-classifier-app</artifactId>
+
+    <packaging>ear</packaging>
+
+    <name>ear-project-it:  SampleEAR-classifier-app Module</name>
+
+    <dependencies>
+        <dependency>
+            <groupId>io.openliberty.tools.it</groupId>
+            <artifactId>SampleEJB-classifier-dep</artifactId>
+            <version>1.0-SNAPSHOT</version>
+            <type>ejb</type>
+            <classifier>the-ejb</classifier>
+        </dependency>
+        <dependency>
+            <groupId>io.openliberty.tools.it</groupId>
+            <artifactId>SampleWAR</artifactId>
+            <version>1.0-SNAPSHOT</version>
+            <type>war</type>
+        </dependency>
+      <dependency>
+            <groupId>commons-httpclient</groupId>
+            <artifactId>commons-httpclient</artifactId>
+            <version>3.1</version>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>junit</groupId>
+            <artifactId>junit</artifactId>
+        </dependency>
+    </dependencies>
+
+    <build>
+        <finalName>${project.artifactId}</finalName>
+        <plugins>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-ear-plugin</artifactId>
+                <version>3.0.2</version>
+                <configuration>
+                    <!-- Tell Maven we are using Java EE 7 -->
+                    <version>7</version>
+                    <!-- Use Java EE ear libraries as needed. Java EE ear 
+                        libraries are in easy way to package any libraries needed in the ear, and 
+                        automatically have any modules (EJB-JARs and WARs) use them -->
+                    <classifier>the-ear</classifier>
+                    <defaultLibBundleDir>lib</defaultLibBundleDir>
+                    <modules>
+                        <webModule>
+                            <groupId>io.openliberty.tools.it</groupId>
+                            <artifactId>SampleWAR</artifactId>
+                            <uri>/modules/web.war</uri>
+                        </webModule>
+                    </modules>
+                </configuration>
+            </plugin>
+            <plugin>
+                <groupId>io.openliberty.tools</groupId>
+                <artifactId>liberty-maven-plugin</artifactId>
+                <version>@pom.version@</version>
+                <configuration>
+                    <stripVersion>true</stripVersion>
+                    <assemblyArtifact>
+                        <groupId>${runtimeGroupId}</groupId>
+                        <artifactId>${runtimeArtifactId}</artifactId>
+                        <version>${runtimeVersion}</version>
+                        <type>zip</type>
+                    </assemblyArtifact>
+                    <serverName>test</serverName>
+                    <appsDirectory>apps</appsDirectory>
+                    <looseApplication>true</looseApplication>
+                    <include>usr</include>
+                </configuration>
+                <executions>
+                    <execution>
+                        <id>create-server</id>
+                        <phase>prepare-package</phase>
+                        <goals>
+                            <goal>install-server</goal>
+                            <goal>create</goal>
+                        </goals>
+                    </execution>
+                    <execution>
+                        <id>install-apps</id>
+                        <phase>package</phase>
+                        <goals>
+                            <goal>deploy</goal>
+                        </goals>
+                    </execution>
+                    <execution>
+                        <id>test-start-server</id>
+                        <phase>pre-integration-test</phase>
+                        <goals>
+                            <goal>test-start</goal>
+                        </goals>
+                    </execution>
+                    <execution>
+                        <id>test-stop-server</id>
+                        <phase>post-integration-test</phase>
+                        <goals>
+                            <goal>test-stop</goal>
+                        </goals>
+                    </execution>
+                </executions>
+            </plugin>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-compiler-plugin</artifactId>
+                <executions>
+                    <execution>
+                        <goals>
+                            <goal>testCompile</goal>
+                        </goals>
+                    </execution>
+                </executions>
+            </plugin>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-failsafe-plugin</artifactId>
+                <version>2.19.1</version>
+                <configuration>
+                    <redirectTestOutputToFile>true</redirectTestOutputToFile>
+                    <forkMode>once</forkMode>
+                    <forkedProcessTimeoutInSeconds>300</forkedProcessTimeoutInSeconds>
+                    <argLine>-enableassertions</argLine>
+                    <workingDirectory>${project.build.directory}</workingDirectory>
+                </configuration>
+                <executions>
+                    <execution>
+                        <goals>
+                            <goal>integration-test</goal>
+                            <goal>verify</goal>
+                        </goals>
+                    </execution>
+                </executions>
+            </plugin>
+        </plugins>
+    </build>
+</project>

--- a/liberty-maven-plugin/src/it/ear-project-it/SampleEAR-classifier-app/src/main/application/META-INF/license
+++ b/liberty-maven-plugin/src/it/ear-project-it/SampleEAR-classifier-app/src/main/application/META-INF/license
@@ -1,0 +1,15 @@
+/*******************************************************************************
+ * (c) Copyright IBM Corporation 2017.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *******************************************************************************/

--- a/liberty-maven-plugin/src/it/ear-project-it/SampleEAR-classifier-app/src/main/liberty/config/server.xml
+++ b/liberty-maven-plugin/src/it/ear-project-it/SampleEAR-classifier-app/src/main/liberty/config/server.xml
@@ -1,0 +1,19 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<server description="jee7sample test server">
+
+    <!-- Enable features -->
+    <featureManager>
+        <feature>javaee-7.0</feature>
+    </featureManager>
+        
+    <keyStore password="keyPassword"/>
+    <basicRegistry id="basic" realm="BasicRealm">
+        <user name="admin" password="password" />
+    </basicRegistry>
+
+    <!-- To access this server from a remote client add a host attribute to the following element, e.g. host="*" -->
+    <httpEndpoint id="defaultHttpEndpoint"
+                  httpPort="9080"
+                  httpsPort="9443" />
+
+</server>

--- a/liberty-maven-plugin/src/it/ear-project-it/SampleEAR-classifier-app/src/test/java/net/wasdev/wlp/maven/test/it/EndpointIT.java
+++ b/liberty-maven-plugin/src/it/ear-project-it/SampleEAR-classifier-app/src/test/java/net/wasdev/wlp/maven/test/it/EndpointIT.java
@@ -1,0 +1,51 @@
+/*******************************************************************************
+ * (c) Copyright IBM Corporation 2017.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *******************************************************************************/
+package net.wasdev.wlp.maven.test.it;
+
+import static org.junit.Assert.*;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.apache.commons.httpclient.HttpClient;
+import org.apache.commons.httpclient.HttpStatus;
+import org.apache.commons.httpclient.methods.GetMethod;
+
+public class EndpointIT {
+    private static String URL;
+    
+    @BeforeClass
+    public static void init() {
+        URL = "http://localhost:9080/SampleWAR/";
+    }
+    
+    @Test
+    public void testServlet() throws Exception {
+        HttpClient client = new HttpClient();
+        
+        GetMethod method = new GetMethod(URL);
+        
+        try {
+            int statusCode = client.executeMethod(method);
+            
+            assertEquals("HTTP GET failed", HttpStatus.SC_OK, statusCode);
+            
+            String response = method.getResponseBodyAsString(1000);
+            
+            assertTrue("Unexpected response body", response.contains("Hello classifier test World."));
+        } finally {
+            method.releaseConnection();
+        }
+    }
+}

--- a/liberty-maven-plugin/src/it/ear-project-it/SampleEAR-classifier-app/src/test/java/net/wasdev/wlp/maven/test/it/LooseConfigTestIT.java
+++ b/liberty-maven-plugin/src/it/ear-project-it/SampleEAR-classifier-app/src/test/java/net/wasdev/wlp/maven/test/it/LooseConfigTestIT.java
@@ -1,0 +1,91 @@
+/*******************************************************************************
+ * (c) Copyright IBM Corporation 2017, 2019.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *******************************************************************************/
+package net.wasdev.wlp.maven.test.it;
+
+import java.io.File;
+import java.io.FileInputStream;
+
+import javax.xml.parsers.DocumentBuilder;
+import javax.xml.parsers.DocumentBuilderFactory;
+import javax.xml.xpath.XPath;
+import javax.xml.xpath.XPathConstants;
+import javax.xml.xpath.XPathFactory;
+
+import org.junit.Test;
+import org.w3c.dom.Document;
+import org.w3c.dom.NodeList;
+import org.w3c.dom.Document;
+
+import static junit.framework.Assert.*;
+
+public class LooseConfigTestIT {
+    
+    public final String LOOSE_APP = "liberty/wlp/usr/servers/test/apps/SampleEAR-classifier-app-the-ear.ear.xml";
+        
+    @Test
+    public void testLooseApplicationFileExist() throws Exception {
+        File f = new File(LOOSE_APP);
+        assertTrue(f.getCanonicalFile() + " doesn't exist", f.exists());
+    }
+    
+    @Test
+    public void testLooseApplicationFileContent() throws Exception {
+        File f = new File(LOOSE_APP);
+        FileInputStream input = new FileInputStream(f);
+        
+        // get input XML Document 
+        DocumentBuilderFactory inputBuilderFactory = DocumentBuilderFactory.newInstance();
+        inputBuilderFactory.setIgnoringComments(true);
+        inputBuilderFactory.setCoalescing(true);
+        inputBuilderFactory.setIgnoringElementContentWhitespace(true);
+        inputBuilderFactory.setValidating(false);
+        DocumentBuilder inputBuilder = inputBuilderFactory.newDocumentBuilder();
+        Document inputDoc=inputBuilder.parse(input);
+        
+        // parse input XML Document
+        XPath xPath = XPathFactory.newInstance().newXPath();
+        String expression = "/archive/file";
+        NodeList nodes = (NodeList) xPath.compile(expression).evaluate(inputDoc, XPathConstants.NODESET);
+        assertEquals("Number of <file/> element ==>", 2, nodes.getLength());
+        assertEquals("file targetInArchive attribute value", "/META-INF/application.xml", 
+                nodes.item(0).getAttributes().getNamedItem("targetInArchive").getNodeValue());
+        assertEquals("file targetInArchive attribute value", "/META-INF/MANIFEST.MF", 
+                nodes.item(1).getAttributes().getNamedItem("targetInArchive").getNodeValue());
+        
+        expression = "/archive/dir";
+        nodes = (NodeList) xPath.compile(expression).evaluate(inputDoc, XPathConstants.NODESET);
+        assertEquals("Number of <dir/> element ==>", 1, nodes.getLength());
+        
+        expression = "/archive/archive";
+        nodes = (NodeList) xPath.compile(expression).evaluate(inputDoc, XPathConstants.NODESET);
+        assertEquals("Number of <archive/> element ==>", 2, nodes.getLength());
+        assertEquals("archive targetInArchive attribute value", "/io.openliberty.tools.it-SampleEJB-classifier-dep-1.0-SNAPSHOT-the-ejb.jar", 
+                nodes.item(0).getAttributes().getNamedItem("targetInArchive").getNodeValue());
+        assertEquals("archive targetInArchive attribute value", "/modules/web.war", 
+                nodes.item(1).getAttributes().getNamedItem("targetInArchive").getNodeValue());
+        
+        expression = "/archive/archive/file";
+        nodes = (NodeList) xPath.compile(expression).evaluate(inputDoc, XPathConstants.NODESET);
+        // Note: log4j-core also pulls in log4j-api
+        assertEquals("Number of <archive/> element ==>", 4, nodes.getLength());
+
+        // test runtime scope dependency to be incldued in the ?WEB-INF/lib
+        assertEquals("file targetInArchive attribute value", "/WEB-INF/lib/log4j-core-2.13.0.jar", 
+                nodes.item(2).getAttributes().getNamedItem("targetInArchive").getNodeValue());
+        assertEquals("file targetInArchive attribute value", "/WEB-INF/lib/log4j-api-2.13.0.jar", 
+                nodes.item(3).getAttributes().getNamedItem("targetInArchive").getNodeValue());
+    }
+}

--- a/liberty-maven-plugin/src/it/ear-project-it/SampleEJB-classifier-dep/pom.xml
+++ b/liberty-maven-plugin/src/it/ear-project-it/SampleEJB-classifier-dep/pom.xml
@@ -1,0 +1,44 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+
+    <parent>
+        <groupId>io.openliberty.tools.it</groupId>
+        <artifactId>ear-project-it</artifactId>
+        <version>1.0-SNAPSHOT</version>
+    </parent>
+
+    <artifactId>SampleEJB-classifier-dep</artifactId>
+    <packaging>ejb</packaging>
+
+    <name>ear-project-it:  SampleEJB-classifier-dep Module</name>
+
+    <dependencies>
+        <dependency>
+            <groupId>javax</groupId>
+            <artifactId>javaee-api</artifactId>
+        </dependency>
+
+        <!-- Test scope dependencies -->
+        <dependency>
+            <groupId>junit</groupId>
+            <artifactId>junit</artifactId>
+        </dependency>
+    </dependencies>
+
+    <build>
+        <finalName>${project.artifactId}</finalName>
+        <plugins>
+            <plugin>
+                <artifactId>maven-ejb-plugin</artifactId>
+                <configuration>
+                    <!-- Tell Maven we are using EJB 3.1 -->
+                    <ejbVersion>3.1</ejbVersion>
+                    <classifier>the-ejb</classifier>
+                </configuration>
+            </plugin>
+        </plugins>
+    </build>
+
+</project>

--- a/liberty-maven-plugin/src/it/ear-project-it/SampleEJB-classifier-dep/src/main/java/wasdev/sample/ejb/SampleStatelessBean.java
+++ b/liberty-maven-plugin/src/it/ear-project-it/SampleEJB-classifier-dep/src/main/java/wasdev/sample/ejb/SampleStatelessBean.java
@@ -1,0 +1,26 @@
+/*******************************************************************************
+ * (c) Copyright IBM Corporation 2017, 2021.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *******************************************************************************/
+package wasdev.sample.ejb;
+
+import javax.ejb.Stateless;
+
+@Stateless
+public class SampleStatelessBean {
+
+    public String hello() {
+        return "Hello classifier test World.";
+    }
+}

--- a/liberty-maven-plugin/src/it/ear-project-it/pom.xml
+++ b/liberty-maven-plugin/src/it/ear-project-it/pom.xml
@@ -20,6 +20,8 @@
         <module>SampleEar_NonLooseApp</module>
         <module>SampleEAR</module>
         <module>SampleWLP</module>
+        <module>SampleEAR-classifier-app</module>
+        <module>SampleEJB-classifier-dep</module>
         <module>skinnywar-ear</module>
         <module>skinnywar-wlp</module>
         <module>helloworld-impl</module>

--- a/liberty-maven-plugin/src/main/java/io/openliberty/tools/maven/applications/DeployMojoSupport.java
+++ b/liberty-maven-plugin/src/main/java/io/openliberty/tools/maven/applications/DeployMojoSupport.java
@@ -188,13 +188,13 @@ public class DeployMojoSupport extends PluginConfigSupport {
                     MavenProject dependencyProject = getReactorMavenProject(artifact);
                     switch (artifact.getType()) {
                     case "jar":
-                        looseEar.addJarModule(dependencyProject);
+                        looseEar.addJarModule(dependencyProject, artifact);
                         break;
                     case "ejb":
-                        looseEar.addEjbModule(dependencyProject);
+                        looseEar.addEjbModule(dependencyProject, artifact);
                         break;
                     case "war":
-                        Element warArchive = looseEar.addWarModule(dependencyProject,
+                        Element warArchive = looseEar.addWarModule(dependencyProject, artifact,
                                 getWarSourceDirectory(dependencyProject));
                         if (looseEar.isEarSkinnyWars()) {
                             // add embedded lib only if they are not a compile dependency in the ear
@@ -205,7 +205,7 @@ public class DeployMojoSupport extends PluginConfigSupport {
                         }
                         break;
                     case "rar":
-                        Element rarArchive = looseEar.addRarModule(dependencyProject);
+                        Element rarArchive = looseEar.addRarModule(dependencyProject, artifact);
                         addEmbeddedLib(rarArchive, dependencyProject, looseEar, "/");
                         break;
                     default:
@@ -288,7 +288,7 @@ public class DeployMojoSupport extends PluginConfigSupport {
         {
             if (isReactorMavenProject(artifact)) {
                 MavenProject dependProject = getReactorMavenProject(artifact);
-                Element archive = looseApp.addArchive(parent, dir + dependProject.getBuild().getFinalName() + ".jar");
+                Element archive = looseApp.addArchive(parent, dir + artifact.getFile().getName());
                 looseApp.addOutputDir(archive, new File(dependProject.getBuild().getOutputDirectory()), "/");
                 
                 File manifestFile = MavenProjectUtil.getManifestFile(dependProject, "maven-jar-plugin");

--- a/liberty-maven-plugin/src/main/java/io/openliberty/tools/maven/applications/DeployMojoSupport.java
+++ b/liberty-maven-plugin/src/main/java/io/openliberty/tools/maven/applications/DeployMojoSupport.java
@@ -257,10 +257,10 @@ public class DeployMojoSupport extends PluginConfigSupport {
         }
     }
 
-    private void addEmbeddedLib(Element parent, MavenProject proj, LooseApplication looseApp, String dir)
+    private void addEmbeddedLib(Element parent, MavenProject warProject, LooseApplication looseApp, String dir)
             throws Exception {
-        Set<Artifact> artifacts = proj.getArtifacts();
-        log.debug("Number of compile dependencies for " + proj.getArtifactId() + " : " + artifacts.size());
+        Set<Artifact> artifacts = warProject.getArtifacts();
+        log.debug("Number of compile dependencies for " + warProject.getArtifactId() + " : " + artifacts.size());
 
         for (Artifact artifact : artifacts) {
             if (("compile".equals(artifact.getScope()) || "runtime".equals(artifact.getScope()))
@@ -270,9 +270,9 @@ public class DeployMojoSupport extends PluginConfigSupport {
         }
     }
 
-    private void addSkinnyWarLib(Element parent, MavenProject proj, LooseEarApplication looseEar) throws Exception {
-        Set<Artifact> artifacts = proj.getArtifacts();
-        log.debug("Number of compile dependencies for " + proj.getArtifactId() + " : " + artifacts.size());
+    private void addSkinnyWarLib(Element parent, MavenProject warProject, LooseEarApplication looseEar) throws Exception {
+        Set<Artifact> artifacts = warProject.getArtifacts();
+        log.debug("Number of compile dependencies for " + warProject.getArtifactId() + " : " + artifacts.size());
 
         for (Artifact artifact : artifacts) {
             // skip the embedded library if it is included in the lib directory of the ear
@@ -288,7 +288,8 @@ public class DeployMojoSupport extends PluginConfigSupport {
         {
             if (isReactorMavenProject(artifact)) {
                 MavenProject dependProject = getReactorMavenProject(artifact);
-                Element archive = looseApp.addArchive(parent, dir + artifact.getFile().getName());
+                String artifactFileName = getPreDeployAppFileName(dependProject);
+                Element archive = looseApp.addArchive(parent, dir + artifactFileName);
                 looseApp.addOutputDir(archive, new File(dependProject.getBuild().getOutputDirectory()), "/");
                 
                 File manifestFile = MavenProjectUtil.getManifestFile(dependProject, "maven-jar-plugin");

--- a/liberty-maven-plugin/src/main/java/io/openliberty/tools/maven/applications/LooseEarApplication.java
+++ b/liberty-maven-plugin/src/main/java/io/openliberty/tools/maven/applications/LooseEarApplication.java
@@ -63,17 +63,17 @@ public class LooseEarApplication extends LooseApplication {
         }
     }
 
-    public Element addJarModule(MavenProject proj) throws Exception {
-        return addModule(proj, "maven-jar-plugin");
+    public Element addJarModule(MavenProject proj, Artifact artifact) throws Exception {
+        return addModule(proj, artifact, "maven-jar-plugin");
     }
 
-    public Element addEjbModule(MavenProject proj) throws Exception {
-        return addModule(proj, "maven-ejb-plugin");
+    public Element addEjbModule(MavenProject proj, Artifact artifact) throws Exception {
+        return addModule(proj, artifact, "maven-ejb-plugin");
     }
 
-    public Element addModule(MavenProject proj, String pluginId) throws Exception {
+    public Element addModule(MavenProject proj, Artifact artifact, String pluginId) throws Exception {
         File outputDirectory = new File(proj.getBuild().getOutputDirectory());
-        Element moduleArchive = config.addArchive(getModuleUri(proj));
+        Element moduleArchive = config.addArchive(getModuleUri(artifact));
         config.addDir(moduleArchive, outputDirectory, "/");
         // add manifest.mf
         File manifestFile = MavenProjectUtil.getManifestFile(proj, pluginId);
@@ -83,17 +83,17 @@ public class LooseEarApplication extends LooseApplication {
         return moduleArchive;
     }
 
-    public Element addWarModule(MavenProject proj, File warSourceDir) throws Exception {
-        Element warArchive = config.addArchive(getModuleUri(proj));
+    public Element addWarModule(MavenProject proj, Artifact artifact, File warSourceDir) throws Exception {
+        Element warArchive = config.addArchive(getModuleUri(artifact));
         config.addDir(warArchive, warSourceDir, "/");
         config.addDir(warArchive, new File(proj.getBuild().getOutputDirectory()), "/WEB-INF/classes");
         // add Manifest file
-        addWarManifestFile(warArchive, proj);
+        addWarManifestFile(warArchive, artifact, proj);
         return warArchive;
     }
 
-    public Element addRarModule(MavenProject proj) throws Exception {
-        Element rarArchive = config.addArchive(getModuleUri(proj));
+    public Element addRarModule(MavenProject proj, Artifact artifact) throws Exception {
+        Element rarArchive = config.addArchive(getModuleUri(artifact));
         config.addDir(rarArchive, getRarSourceDirectory(proj), "/");
 
         // get raXmlFile optional rar plugin parameter
@@ -118,10 +118,6 @@ public class LooseEarApplication extends LooseApplication {
         } else {
             return new File(proj.getBasedir(), "src/main/rar");
         }
-    }
-
-    public String getModuleUri(MavenProject proj) throws Exception {
-        return getModuleUri(proj.getArtifact());
     }
 
     public String getModuleUri(Artifact artifact) throws Exception {
@@ -272,10 +268,10 @@ public class LooseEarApplication extends LooseApplication {
         }
     }
 
-    public void addWarManifestFile(Element parent, MavenProject proj) throws Exception {
+    public void addWarManifestFile(Element parent, Artifact artifact, MavenProject proj) throws Exception {
         // the ear plug-in modify the skinnyWar module manifest file in
         // ${project.build.directory}/temp
-        File newMf = new File(project.getBuild().getDirectory() + "/temp/" + getModuleUri(proj) + "/META-INF");
+        File newMf = new File(project.getBuild().getDirectory() + "/temp/" + getModuleUri(artifact) + "/META-INF");
         if (isEarSkinnyWars() && newMf.exists()) {
             config.addDir(parent, newMf, "/META-INF");
         } else {

--- a/liberty-maven-plugin/src/main/java/io/openliberty/tools/maven/utils/MavenProjectUtil.java
+++ b/liberty-maven-plugin/src/main/java/io/openliberty/tools/maven/utils/MavenProjectUtil.java
@@ -99,7 +99,9 @@ public class MavenProjectUtil {
     public static String getAppNameClassifier(MavenProject proj) {
         String pluginName = null;
 
-        if (proj.getPackaging().equals("war")) {
+        if (proj.getPackaging().equals("jar")) {
+            pluginName = "maven-jar-plugin";
+        } else if (proj.getPackaging().equals("war")) {
             pluginName = "maven-war-plugin";
         } else if (proj.getPackaging().equals("ear")) {
             pluginName = "maven-ear-plugin";

